### ISSUE TITLE
Implement `flake8-no-implicit-concat`

### ISSUE THREAD

**Issue by ngnpope** [charliermarsh/ruff#2362]
[GitHub](https://github.com/10sr/flake8-no-implicit-concat), [PyPI](https://pypi.org/project/flake8-no-implicit-concat/).

- [x] [`NIC001`](https://github.com/10sr/flake8-no-implicit-concat#violation-codes): Implicitly concatenated str literals on one line
  - Already implemented via `ISC001` from [`flake8-implicit-str-concat`](https://github.com/charliermarsh/ruff#flake8-implicit-str-concat-isc).
- [x] [`NIC002`](https://github.com/10sr/flake8-no-implicit-concat#violation-codes): Implicitly concatenated str literals over multiple lines
  - Already implemented via `ISC002` from [`flake8-implicit-str-concat`](https://github.com/charliermarsh/ruff#flake8-implicit-str-concat-isc) when using the configuration option provided by `ruff` (`allow-multiline = false`)
- [x] [`NIC101`](https://github.com/10sr/flake8-no-implicit-concat#violation-codes): Implicitly concatenated bytes literals on one line
  - Already implemented via `ISC001` from [`flake8-implicit-str-concat`](https://github.com/charliermarsh/ruff#flake8-implicit-str-concat-isc).
- [x] [`NIC102`](https://github.com/10sr/flake8-no-implicit-concat#violation-codes): Implicitly concatenated bytes literals over multiple lines
  - Already implemented via `ISC002` from [`flake8-implicit-str-concat`](https://github.com/charliermarsh/ruff#flake8-implicit-str-concat-isc) when using the configuration option provided by `ruff` (`allow-multiline = false`)

This is somewhat the inverse of [`flake8-implicit-str-concat`](https://github.com/charliermarsh/ruff#flake8-implicit-str-concat-isc). Despite this, they agree that implicit concatenation on a single line shouldn't happen. One also has a separate rule for `bytes` vs `str` where the other uses combined rules.

**Comment by charliermarsh** [comment]
I think we actually _do_ support `NIC002` via [this setting](https://github.com/charliermarsh/ruff#flake8-implicit-str-concat), which defaults to `true` for compatibility with `flake8-implicit-str-concat`.

**Comment by ngnpope** [comment]
Ah, I think it's not quite the same... Lemme check.

**Comment by ngnpope** [comment]
Nope. You're right 😄 I missed that the code was different...

We probably want a documentation bit to state that setting `allow-multiline = false` for `ISC002` requires disabling `ISC003` to prevent them conflicting. (Or we could disable `ISC003`?)

```python
implicit_single_line = "abc" "xyz"

explicit_single_line = "abc" + "xyz"


implicit_multiple_line_continuation = "abc" \
    "xyz"

explicit_multiple_line_continuation_after = "abc" \
    + "xyz"

explicit_multiple_line_continuation_before = "abc" + \
    "xyz"


implicit_multiple_line_parentheses = (
    "abc"
    "xyz"
)

explicit_multiple_line_parentheses_after = (
    "abc" +
    "xyz"
)

explicit_multiple_line_parentheses_before = (
    "abc"
    + "xyz"
)
```

For which I got the following outputs:

```
### ruff w/ allow-multiline = true
# example.py:1:24: ISC001 Implicitly concatenated string literals on one line
# example.py:3:24: ISC003 Explicitly concatenated string should be implicitly concatenated
# example.py:6:39: ISC002 Implicitly concatenated string literals over continuation line
# example.py:9:45: ISC003 Explicitly concatenated string should be implicitly concatenated
# example.py:12:46: ISC003 Explicitly concatenated string should be implicitly concatenated
# example.py:22:5: ISC003 Explicitly concatenated string should be implicitly concatenated
# example.py:27:5: ISC003 Explicitly concatenated string should be implicitly concatenated

### ruff w/ allow-multiline = false
# example.py:1:24: ISC001 Implicitly concatenated string literals on one line
# example.py:3:24: ISC003 Explicitly concatenated string should be implicitly concatenated
# example.py:6:39: ISC002 Implicitly concatenated string literals over continuation line
# example.py:9:45: ISC003 Explicitly concatenated string should be implicitly concatenated
# example.py:12:46: ISC003 Explicitly concatenated string should be implicitly concatenated
# example.py:17:5: ISC002 Implicitly concatenated string literals over continuation line
# example.py:22:5: ISC003 Explicitly concatenated string should be implicitly concatenated
# example.py:27:5: ISC003 Explicitly concatenated string should be implicitly concatenated

### flake8-implicit-str-concat
# example.py:1:29: ISC001 implicitly concatenated string literals on one line
# example.py:3:24: ISC003 explicitly concatenated string should be implicitly concatenated
# example.py:6:44: ISC002 implicitly concatenated string literals over continuation line
# example.py:9:45: ISC003 explicitly concatenated string should be implicitly concatenated
# example.py:12:46: ISC003 explicitly concatenated string should be implicitly concatenated
# example.py:22:5: ISC003 explicitly concatenated string should be implicitly concatenated
# example.py:27:5: ISC003 explicitly concatenated string should be implicitly concatenated

### flake8-no-implicit-concat
# example.py:1:29: NIC001 Implicitly concatenated str literals on one line
# example.py:6:44: NIC002 Implicitly concatenated str literals over multiple lines
# example.py:17:10: NIC002 Implicitly concatenated str literals over multiple lines
```

**Comment by ngnpope** [comment]
Well, other than the potential disabling of `ISC003` if `allow-multiline = false`, we can chalk this up as another supported plugin.
